From 3aec93e6f646e58fc255ef135916d554848bfe32 Mon Sep 17 00:00:00 2001
From: PulkoMandy <pulkomandy@pulkomandy.tk>
Date: Sat, 21 Jun 2025 20:21:15 +0200
Subject: Get it to build on Haiku

- Disable search for ucontext.h on Haiku (Haiku defines ucontext struct in signal.h)
- Link libsd (for strsep) and libnetwork (for sockets)
- Disable code for TUN/TAP handling (Haiku specific code needs to be
  written for this): _IOW and _IOR don't exist on Haiku, and the TUN configration uses a different
  interface.
- Use elf_i386_haiku instead of elf_i386 (not sure why Haiku's binutils don't know about elf_i386)
- Implement CPU features detection (there is no /proc/cpuinfo in Haiku; use cpuid and
  get_cpu_topology_info instead)
- Missing include of sys/select.h for use of fd_set (required by POSIX but not needed on some other
  OS)
- Enable definition of getmcontext and setmcontext for Haiku
- In Haiku, the libexec directory is /system/lib and the lib directory is /system/develop/lib. This
  should be handled using LIBDIR and LIBEXECDIR ideally (just like DATADIR and other directories
  handled by autoconf)

diff --git a/configure.ac b/configure.ac
index 25da201..5a6aa6a 100644
--- a/configure.ac
+++ b/configure.ac
@@ -501,6 +501,12 @@ AC_CHECK_HEADERS([asm/ucontext.h],,, [
   #include <ucontext.h>
   #include <signal.h>
 ])
+
+if test "$CONFIG_OS" == "Haiku"; then
+    LIBS="$LIBS -lbsd -lnetwork"
+fi
+
+if test "$CONFIG_OS" != "Haiku"; then
 if test "$machine" != "wasm32"; then
   AC_CHECK_HEADERS([ucontext.h],, [
     AS_UNSET([ac_cv_header_ucontext_h])
@@ -510,6 +516,7 @@ if test "$machine" != "wasm32"; then
     #define _XOPEN_SOURCE 500
   ])])
 fi
+fi
 AC_CHECK_HEADERS([linux/signal.h],,, [#include <signal.h>])
 AC_CHECK_HEADERS([sys/soundcard.h], [
   USE_OSS=1
@@ -840,7 +847,7 @@ fi
 DOSEMU_CFLAGS="${DOSEMU_CFLAGS} ${OPT} ${PIPE}"
 DOSEMU_CPPFLAGS="${DOSEMU_CPPFLAGS} -MD -DCFLAGS_STR=\"$DOSEMU_CFLAGS $CFLAGS\""
 DOSEMU_VERSION=`cd $srcdir && ./getversion -b`
-AS_LDFLAGS="-melf_i386"
+AS_LDFLAGS="-melf_i386_haiku"
 
 AC_SUBST(XASFLAGS)
 AC_SUBST(ASFLAGS)
diff --git a/src/base/init/config.c b/src/base/init/config.c
index efda100..668eaa2 100644
--- a/src/base/init/config.c
+++ b/src/base/init/config.c
@@ -43,6 +43,14 @@
 #include "mhpdbg.h"
 #include "mapping/mapping.h"
 
+#ifdef __HAIKU__
+/* Rename some conflicting functions that we don't need */
+#define read_port haiku_read_port
+#define write_port haiku_write_port
+#include <kernel/OS.h>
+#undef read_port
+#undef write_port
+#endif
 /*
  * Options used in config_init().
  *
@@ -812,10 +838,84 @@ void secure_option_preparse(int *argc, char **argv)
 
 static void read_cpu_info(void)
 {
-    char *cpuflags, *cpu;
     int k = 3;
+    int cores = 0;
     int err;
 
+#ifdef __HAIKU__
+    cpuid_info info;
+
+    err = get_cpuid(&info, 1, 0);
+    if (err != B_OK)
+        return;
+
+    k = info.eax_1.family;
+    bool fpu = info.eax_1.extended_features & (1 << 0);
+    bool tsc = info.eax_1.extended_features & (1 << 4);
+    bool mmx = info.eax_1.extended_features & (1 << 23);
+    bool fxsr = info.eax_1.extended_features & (1 << 24);
+    bool sse = info.eax_1.extended_features & (1 << 25);
+
+    cpu_topology_node_info topology[128];
+    int count = 128;
+    get_cpu_topology_info(topology, &count);
+
+    if (k >= 5) {
+        config.realcpu = CPU_586;
+
+#ifdef X86_EMULATOR
+        if (mmx || sse) {
+	  config.cpuprefetcht0 = 1;
+        }
+#endif
+#ifdef __i386__
+        if (fxsr) {
+          config.cpufxsr = 1;
+	  if (sse)
+	    config.cpusse = 1;
+	}
+#endif
+        if (tsc) {
+            long long chz = 0;
+            for (int i = 0; i < count; i++) {
+                if (topology[i].type == B_TOPOLOGY_CORE) {
+                    chz = topology[i].data.core.default_frequency;
+                    cores ++;
+                }
+            }
+		/* speed division factor to get 1us from CPU clock */
+		config.cpu_spd = (LLF_US*1000000)/chz;
+
+		/* speed division factor to get 838ns from CPU clock */
+		config.cpu_tick_spd = (LLF_TICKS*1000000)/chz;
+
+		config.CPUSpeedInMhz = chz / 1000000;
+
+        }
+    } else if (k == 4) {
+        config.realcpu = CPU_486;
+    } else if (k != 3) {
+        error("Unknown CPU type!\n");
+    }
+
+    if (config.mathco) {
+        config.mathco = fpu;
+    }
+
+    err = get_cpuid(&info, 9, 0);
+    if (err != B_OK)
+      return;
+
+    if (info.regs.ecx & (1 << 2))
+      config.umip = 1;
+    else
+      warn("Your CPU doesn't support UMIP\n");
+
+    if (cores > 0)
+        config.smp = 1;
+#else
+    char *cpuflags, *cpu;
+
     err = open_proc_scan("/proc/cpuinfo");
     if (err)
       return;
@@ -853,6 +953,7 @@ static void read_cpu_info(void)
 	    config.cpusse = 1;
 	}
 #endif
+
         if (cpuflags && strstr(cpuflags, "tsc")) {
           /* bogospeed currently returns 0; should it deny
            * pentium features, fall back into 486 case */
@@ -921,6 +1022,7 @@ static void read_cpu_info(void)
       config.smp = 1;		/* for checking overrides, later */
     }
     close_proc_scan();
+#endif
 }
 
 static void config_post_process(void)
diff --git a/src/base/lib/mcontext/asm.SS b/src/base/lib/mcontext/asm.SS
index fefe877..bf1dba5 100644
--- a/src/base/lib/mcontext/asm.SS
+++ b/src/base/lib/mcontext/asm.SS
@@ -12,6 +12,18 @@
 #endif
 #endif
 
+#if defined(__HAIKU__) && defined(__i386__)
+#define NEEDX86CONTEXT 1
+#define SET setmcontext
+#define GET getmcontext
+#else
+#if defined(__HAIKU__) && defined(__x86_64__)
+#define NEEDAMD64CONTEXT 1
+#define SET _setmcontext
+#define GET _getmcontext
+#endif
+#endif
+
 #if defined(__OpenBSD__) && defined(__i386__)
 #define NEEDX86CONTEXT 1
 #define SET setmcontext
diff --git a/src/base/misc/utilities.c b/src/base/misc/utilities.c
index 087561a..b2f4774 100644
--- a/src/base/misc/utilities.c
+++ b/src/base/misc/utilities.c
@@ -598,7 +598,7 @@ char *path_prefix(const char *suffix)
         ret = assemble_path(p, suffix);
       else if (strcmp(p1 + 1, "dosemu2") == 0) {
         p1 = strrchr(p, '/');
-        if (p1 && strcmp(p1 + 1, "libexec") == 0) {
+        if (p1 && strcmp(p1 + 1, "lib") == 0) {
           *p1 = '\0';
           ret = assemble_path(p, suffix);
         }
diff --git a/src/dosemu b/src/dosemu
index d423af5..176f715 100644
--- a/src/dosemu
+++ b/src/dosemu
@@ -23,7 +23,7 @@ get_binary() {
     fi
 
   else
-    BINARY="$prefix"/libexec/dosemu2/dosemu2.bin
+    BINARY="$prefix"/lib/dosemu2/dosemu2.bin
     if [ ! -f "$BINARY" ]; then
       echo "$BINARY does not exist"
       exit 1
diff --git a/src/dosext/net/libpacket.c b/src/dosext/net/libpacket.c
index a2a359f..66130f9 100644
--- a/src/dosext/net/libpacket.c
+++ b/src/dosext/net/libpacket.c
@@ -423,6 +423,9 @@ int GetDeviceMTU(void)
 
 int tun_alloc(char *dev)
 {
+#ifdef __HAIKU__
+      return -1;
+#else
       PRIV_SAVE_AREA
       struct ifreq ifr;
       int fd, err;
@@ -459,6 +462,7 @@ int tun_alloc(char *dev)
       strcpy(dev, ifr.ifr_name);
 
       return fd;
+#endif
 }
 
 static ssize_t pkt_read_eth(int pkt_fd, void *buf, size_t count)
diff --git a/src/include/Linux/if_tun.h b/src/include/Linux/if_tun.h
index 5feacf5..16f9184 100644
--- a/src/include/Linux/if_tun.h
+++ b/src/include/Linux/if_tun.h
@@ -26,6 +26,7 @@
 #define TUN_TYPE_MASK   0x000f
 
 /* Ioctl defines */
+#ifndef __HAIKU__
 #define TUNSETNOCSUM  _IOW('T', 200, int) 
 #define TUNSETDEBUG   _IOW('T', 201, int) 
 #define TUNSETIFF     _IOW('T', 202, int) 
@@ -54,6 +55,7 @@
  */
 #define TUNSETVNETBE _IOW('T', 222, int)
 #define TUNGETVNETBE _IOR('T', 223, int)
+#endif
 
 /* TUNSETIFF ifr flags */
 #define IFF_TUN		0x0001
diff --git a/src/include/dosemu_config.h b/src/include/dosemu_config.h
index 0b18a0d..bbd7a8c 100644
--- a/src/include/dosemu_config.h
+++ b/src/include/dosemu_config.h
@@ -53,7 +53,11 @@ extern int config_check_only;
 #define  FREEDOS_DIR        "freedos"         /* freedos dir name */
 #define  FDBOOT_DIR         "fdboot"          /* freedos boot dir name */
 #define  XBAT_DIR           "bat"             /* extras */
+#ifdef __HAIKU__
+#define  DOSEMULIB_DEFAULT  "data/dosemu"
+#else
 #define  DOSEMULIB_DEFAULT  "share/dosemu"
+#endif
 #define  DOSEMUCMDS_DEFAULT DOSEMULIB_DEFAULT "/" CMDS_SUFF
 #define  DOSEMUEXEC_DEFAULT  LIBEXECDIR "/dosemu"
 #define  DOSEMUIMAGE_DEFAULT "~/" LOCALDIR_BASE_NAME
diff --git a/src/plugin/midimisc/mid_i_pipe.c b/src/plugin/midimisc/mid_i_pipe.c
index c54a45e..6b4d9e5 100644
--- a/src/plugin/midimisc/mid_i_pipe.c
+++ b/src/plugin/midimisc/mid_i_pipe.c
@@ -26,6 +26,7 @@
 #include "init.h"
 #include "sound/midi.h"
 #include <sys/types.h>
+#include <sys/select.h>
 #include <sys/stat.h>
 #include <unistd.h>
 #include <fcntl.h>
-- 
2.48.1

